lib/ansible/modules/cloud/cloudscale/cloudscale_volume.py

```
@@ -0,0 +1,284 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# (c) 2018, Gaudenz Steinlin <gaudenz.steinlin@cloudscale.ch>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: cloudscale_volume
+short_description: Manages volumes on the cloudscale.ch IaaS service
+description:
+  - Create, attach/detach and delete volumes on the cloudscale.ch IaaS service.
+  - All operations are performed using the cloudscale.ch public API v1.
+  - "For details consult the full API documentation:
+     U(https://www.cloudscale.ch/en/api/v1)."
+  - A valid API token is required for all operations. You can create as many
+    tokens as you like using the cloudscale.ch control panel at
+    U(https://control.cloudscale.ch).
+notes:
+  - Instead of the api_token parameter the CLOUDSCALE_API_TOKEN environment
+    variable can be used.
+  - To create a new volume at least the C(name) and C(size_gb) options
+    are required.
+version_added: 2.8
+author: "Gaudenz Steinlin (@gaudenz) <gaudenz.steinlin@cloudscale.ch>"
+options:
+  state:
+    description:
+      - State of the volume.
+    default: present
+    choices: [ present, absent ]
+  name:
+    description:
+      - Name of the volume. Either name or UUID must be present to change an
+        existing volume.
+  uuid:
+    description:
+      - UUID of the volume. Either name or UUID must be present to change an
+        existing volume.
+  size_gb:
+    description:
+      - Size of the volume in GB.
+  type:
+    description:
+      - Type of the volume. Cannot be changed after creating the volume.
```

it seems to me, the default would be ssd, defaulted by the API, I would document this here, or alternatively, set the default by ansible. 

(Note, usually it is dangerous to default in ansible because a resource once created would unexpectedly change on a later run, if the param would be removed afterwards. But because you wrote that `type` can only be used "on creation" it would be safe here.)

---

I opted for documenting the default value here. The problem with setting the default in the argument_spec is that it's then no longer possible to distinguish if the option was set to "ssd" explicitly or by the default value. But this distinction is needed for the failure case where someone tries to change the type of an existing volume. Alternatives would be either silently ignore this case or to require the "type: bulk" option even when changing an existing volume. IMO both alternatives are worse.